core/src/main/java/org/apache/iceberg/data/FormatModelRegistry.java

```diff
@@ -0,0 +1,234 @@
+ *   <li>{@link EqualityDeleteWriteBuilder} for equality deletes,
+ *   <li>{@link PositionDeleteWriteBuilder} for position deletes.
+ * </ul>
+ *
+ * The appropriate builder is selected based on {@link FileFormat} and object model name.
+ *
+ * <p>{@link FormatModel} objects are registered through {@link #register(FormatModel)} and used for
+ * creating readers and writers. Read builders are returned directly from the factory. Write
+ * builders may be wrapped in specialized content file writer implementations depending on the
+ * requested builder type.
+ */
+public final class FormatModelRegistry {
+  private static final Logger LOG = LoggerFactory.getLogger(FormatModelRegistry.class);
+  // The list of classes which are used for registering the reader and writer builders
+  private static final List<String> CLASSES_TO_REGISTER = ImmutableList.of();
+
+  private static final Map<Pair<FileFormat, String>, FormatModel<?>> FORMAT_MODELS =
+      Maps.newConcurrentMap();
+
+  /**
+   * Registers an {@link FormatModel} in this registry.
+   *
+   * <p>The {@link FormatModel} creates readers and writers for a specific combinations of file
+   * format (Parquet, ORC, Avro) and object model (for example: "generic", "spark", "flink", etc.).
+   * Registering custom factories allows integration of new data processing engines for the
+   * supported file formats with Iceberg's file access mechanisms.
+   *
+   * <p>Each factory must be uniquely identified by its combination of file format and object model
+   * name. This uniqueness constraint prevents ambiguity when selecting factories for read and write
+   * operations.
+   *
+   * @param formatModel the factory implementation to register
+   * @throws IllegalArgumentException if a factory is already registered for the combination of
+   *     {@link FormatModel#format()} and {@link FormatModel#modelName()}
+   */
+  @SuppressWarnings("CatchBlockLogException")
+  public static void register(FormatModel<?> formatModel) {
+    Pair<FileFormat, String> key = Pair.of(formatModel.format(), formatModel.modelName());
+    if (FORMAT_MODELS.containsKey(key)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Object model factory %s clashes with %s. Both serve %s",
+              formatModel.getClass(), FORMAT_MODELS.get(key), key));
+    }
+
+    FORMAT_MODELS.put(key, formatModel);
+  }
+
+  @SuppressWarnings("CatchBlockLogException")
+  private static void registerSupportedFormats() {
+    // Uses dynamic methods to call the `register` for the listed classes
+    for (String classToRegister : CLASSES_TO_REGISTER) {
+      try {
+        DynMethods.builder("register").impl(classToRegister).buildStaticChecked().invoke();
+      } catch (NoSuchMethodException e) {
+        // failing to register a factory is normal and does not require a stack trace
+        LOG.info("Unable to register {} for data files: {}", classToRegister, e.getMessage());
```

Comment:
It looks like this exception message is based on the `InternalData` one. That specified metadata files to give additional context, but this could be for many different things so I'd probably remove the "for data files".